gensim/models/keyedvectors.py

```diff
@@ -1695,6 +1697,64 @@ def intersect_word2vec_format(self, fname, lockf=0.0, binary=False, encoding='ut
             msg=f"merged {overlap_count} vectors into {self.vectors.shape} matrix from {fname}",
         )
 
+    def vectors_for_all(self, keys: Iterable, allow_inference: bool = True,
+                        copy_vecattrs: bool = False) -> 'KeyedVectors':
+        """Produce vectors for all given keys as a new :class:`KeyedVectors` object.
+
+        Notes
+        -----
+        The keys will always be deduplicated. For optimal performance, you should not pass entire
+        corpora to the method. Instead, you should construct a dictionary of unique words in your
+        corpus:
+
+        >>> from collections import Counter
+        >>> import itertools
+        >>> import gensim.downloader as api
+        >>> from gensim.test.utils import common_texts
+        >>>
+        >>> model = api.load('word2vec-google-news-300')
+        >>> corpus = common_texts
+        >>> word_counts = Counter(itertools.chain.from_iterable(corpus))  # count words in your corpus
+        >>> words_by_freq = (k for k, v in word_counts.most_common())
+        >>> word_vectors = model.wv.vectors_for_all(words_by_freq)  # create word-vectors for words in your corpus
+
+        Parameters
+        ----------
+        keys : iterable
+            The keys that will be vectorized.
+        allow_inference : bool, optional
+            In subclasses such as :class:`~gensim.models.fasttext.FastTextKeyedVectors`,
+            vectors for out-of-vocabulary keys (words) may be inferred. Default is True.
+        copy_vecattrs : bool, optional
+            Additional attributes set via the :meth:`KeyedVectors.set_vecattr` method
+            will be preserved in the produced :class:`KeyedVectors` object. Default is False.
+            To ensure that *all* the produced vectors will have vector attributes assigned,
+            you should set `allow_inference=False`.
+
+        Returns
+        -------
+        keyedvectors : :class:`~gensim.models.keyedvectors.KeyedVectors`
+            Vectors for all the given keys.
+
+        """
+        vocab = (key for key in keys if key in (self if allow_inference else self.key_to_index))  # drop undefined keys
```

Comment:
This comprehension feels a little busy to me. I think this would be more readable:

```suggestion
        if allow_inference:
            vocab = list(OrderedDict.fromkeys((key for key in keys if key in self)))
        else:
            vocab = list(OrderedDict.fromkeys((key for key in keys if key in self.key_to_index)))
```

If we're planning to use type checking later, we should avoid redefining variables to be of different type. For example, in the original code, you start with vocab being a generator, but then turn it into a list. Mypy will complain about that. It's better to either use a different variable, or avoid reassignment altogether as I did in the suggestion above.

Reply:
> This comprehension feels a little busy to me. I think this would be more readable:

We had a more readable version in the course of the review: https://github.com/RaRe-Technologies/gensim/pull/3157#discussion_r642614592 I find it preferable to this code, which features heavy duplication.

Reply:
Since we're in the bike shed territory, that whole `OrderedDict` approach seems unnecessarily opaque to me. That's just for deduplication that keeps the keys in the original order, right? Can you remind me, why do we even want that?

And assuming we do:

```python
        # Pick only the keys that actually exist & deduplicate them.
        vocab, seen = [], set()
        for key in keys:
            if key not in seen:
                seen.add(key)
                if key in (self if allow_inference else self.key_to_index):
                    vocab.append(key)
```

And assuming we don't:

```python
# Pick only the keys that actually exist & deduplicate them.
vocab = list(set(key for key in vocab if key in (self if allow_inference else self.key_to_index)))
```

Reply:
@piskvorky To keep high-frequency terms clustered together for cache-warmness, see https://github.com/RaRe-Technologies/gensim/pull/3157#issuecomment-848081847. The first version with a set is nice and readable.

Reply:
>  for cache-warmnes

Is that even measurable? Sounds like a premature optimization to me.

Reply:
@gojomo profiled the code of word2vec and fastText in the past, so I assumed his suggestion is based on this experience.

Reply:
OK, thanks. I have no strong feeling either way, any of the two options (or even the current PR code) is fine with me.

Reply:
It's conventional, but not required, for assignment of words-to-ordered-positions to put the most-frequent words early. This makes it easy to limit operations to the most-common words - for space or time efficiency, or because the long-tail of rare words tend to have poorer-quality vectors, etc. But also: I've seen mention in papers, and observed long-ago in my own tests, that realistic workloads (including training itself) go faster with frequency-sorted vector assignment - by ensuring the most-looked-up/modified vectors are all in the same few always-in-cache pages. 

So, while it's not a hard requirement to always do/maintain such ordering, we should try not to inadvertently make choices that (a) ruin/scramble that beneficial ordering, when it already exists; or (b) override a users' desired ordering, for these benefits, or anything else the user might want to test.

Reply:
Yes, whichever. Just put a better comment there please @Witiko, explaining the rationale (see my example above), so someone doesn't inadvertently remove this later thinking it serves no purpose.

Reply:
Gotcha!

Reply:
Fixed in ebc276d1.